File: src/main/java/xyz/yuanowo/keelungsightsviewer/crawler/SightsCrawler.java
package xyz.yuanowo.keelungsightsviewer.crawler;

import org.jsoup.Jsoup;
import org.jsoup.nodes.Document;
import org.jsoup.nodes.Element;
import org.jsoup.nodes.TextNode;
import org.jsoup.select.Elements;
import xyz.yuanowo.keelungsightsviewer.model.Image;
import xyz.yuanowo.keelungsightsviewer.model.Sight;

import java.io.IOException;
import java.util.ArrayList;
import java.util.List;
import java.util.regex.Matcher;
import java.util.regex.Pattern;


public class SightsCrawler {

    static String DOMAIN = "https://www.travelking.com.tw";
    static String UA =
            "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/137.0.0.0 " +
                    "Safari/537.36";

    public static void main(String[] args) throws IOException {
        SightsCrawler crawler = new SightsCrawler();
        crawler.getSightsList("七堵");
    }

    public List<Sight> getSightsList(String district) throws IOException {
        Document doc = getDocument("/tourguide/taiwan/keelungcity/");

        Element guidePoint = doc.getElementById("guide-point");
        assert guidePoint != null; // 找不到這個元素就GG了

        Elements districtHeader = guidePoint.select("h4:contains(" + district + ")");

        if (districtHeader.isEmpty()) {
            return List.of();
        }

        List<Sight> sights = new ArrayList<>();

        districtHeader.get(0).nextElementSibling().children().forEach(li -> {
            Element a = li.child(0);
            try {
                Sight sight = getSightDetails(a.attr("href"));
                sights.add(sight);
            } catch (IOException ignored) { }

        });

        return sights;
    }

    public Sight getSightDetails(String url) throws IOException {
        Document doc = getDocument(url);

        Sight sight = new Sight();

        sight.setSourceURL(DOMAIN + url);
        sight.setId(getIDFromURL(url));
        sight.setName(selectText(doc, "h1 > span"));
        sight.setDistrict(selectText(doc, ".breadcrumb .bc_last a"));
        sight.setCategory(selectText(doc, ".point_type > span:nth-child(2)"));
        sight.setDescription(getElementText(doc.selectFirst(".text")));

        Element addressElem = doc.selectFirst(".address a");
        if (addressElem != null) {
            sight.setMapURL(addressElem.attr("href"));
            sight.setAddress(addressElem.text());
        }

        Element img = doc.selectFirst("#galleria .swiper-slide img");
        if (img != null) {
//            Image image = new Image();
//            image.setUrl(img.attr("data-src"));
//            image.setAlternateText(img.attr("alt"));
//            sight.setPhoto(image);
            sight.setPhotoURL(img.attr("data-src"));
        }

        return sight;
    }

    private Long getIDFromURL(String url) {
        Matcher m = Pattern.compile("/tourguide/scenery(\\d+)\\.html").matcher(url);
        if (m.find()) {
            return Long.parseLong(m.group(1));
        } else {
            return null;
        }
    }

    private Document getDocument(String url) throws IOException {
        System.out.println("Fetching " + url);
        Document doc = Jsoup.connect(DOMAIN + url)
                .userAgent(UA)
                .timeout(10 * 1000)
                .get();
        System.out.println("Fetched " + doc.title());
        return doc;
    }

    private String selectText(Document doc, String cssQuery) {
        Element elem = doc.selectFirst(cssQuery);
        if (elem == null) {
            throw new RuntimeException("No such element: " + cssQuery);
        }
        return elem.text();
    }

    private String getElementText(Element elem) {
        if (elem == null) return "";

        StringBuilder builder = new StringBuilder();
        elem.childNodes().forEach(e -> {
            if (e instanceof TextNode el) { // 文字節點
                if (el.isBlank()) return; // 忽略空白
                builder.append(el.text().strip());
            } else if (e instanceof Element el) {
                String text = el.text().strip();
                if (el.tagName().equals("br")) {
                    // 換行符號
                    builder.append("\n");
                } else if (el.tagName().equals("h4")) {
                    // H4 小標題
                    builder.append("#### ").append(text).append("\n");
                } else if (el.tagName().equals("a") && !el.id().equals("po")) {
                    // 連結
                    builder.append(text);
                } else if (el.tagName().equals("strong")) {
                    // 粗體
                    builder.append("**").append(text).append("**");
                } else if (el.className().equals("othermsg")) {
                    builder.append(getElementText(el));
                } else if (!(el.classNames().contains("author") || el.classNames().contains("clear") || el.id().equals("po"))) {
                    System.out.println("Unknown node: " + e);
                }
            } else {
                System.out.println("Unknown node: " + e);
            }
        });
        return builder.toString().strip();
    }

}

File: src/main/java/xyz/yuanowo/keelungsightsviewer/crawler/SightsCrawler.kt
package xyz.yuanowo.keelungsightsviewer.crawler

import kotlinx.coroutines.*
import okhttp3.ConnectionPool
import okhttp3.OkHttpClient
import okhttp3.Request
import org.jsoup.Jsoup
import org.jsoup.nodes.Document
import org.slf4j.LoggerFactory
import xyz.yuanowo.keelungsightsviewer.model.Sight
import java.net.SocketTimeoutException
import java.util.concurrent.TimeUnit

fun main() {
    println("Hello, SightsCrawler!")
    val crawler = SightsCrawler()

    val sights = crawler.fetchSightsListSync()
    println("Found ${sights.size} sights in total")

//    for (district in listOf("七堵", "中山", "中正", "仁愛", "安樂", "信義", "暖暖", "大安", "大安鄉")) {
//        val sights = crawler.fetchSightsListSync(district)
//        println("Found ${sights.size} sights in $district")
//    }
}

val pool = ConnectionPool(10, 3, TimeUnit.MINUTES)

val client = OkHttpClient.Builder()
    .connectionPool(pool)
    .connectTimeout(5, TimeUnit.SECONDS)
    .readTimeout(10, TimeUnit.SECONDS)
    .build()

@OptIn(ExperimentalCoroutinesApi::class)
val limitedIO = Dispatchers.IO.limitedParallelism(10)

class SightsCrawler {

    companion object {

        private const val BASE_URL = "https://www.travelking.com.tw"
        private const val UA =
            "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/137.0.0.0"
        private val log = LoggerFactory.getLogger(SightsCrawler::class.java)

    }

    private suspend fun fetchHtml(url: String): Document? = withContext(limitedIO) {
        log.info("Fetching URL {}", url)
        val request = Request.Builder().url(BASE_URL + url).header("User-Agent", UA).build()
        client.newCall(request).execute().use { response ->
            log.info("Fetched URL {} (Status: {})", url, response.code)
            if (response.isSuccessful) response.body?.let { Jsoup.parse(it.string()) }
            else null
        }
    }

    private suspend fun fetchHtmlWithRetry(url: String, retries: Int = 3, delayMillis: Long = 1000): Document? {
        var waitTime = delayMillis
        repeat(retries) { attempt ->
            try {
                return fetchHtml(url)
            } catch (ex: SocketTimeoutException) {
                if (attempt == retries - 1) {
                    log.error("Failed to fetch {} after {} attempts", url, retries, ex)
                } else {
                    log.warn("Timeout when fetching {} (attempt {}/{})", url, attempt + 1, retries)
                }
                delay(waitTime)
                waitTime *= 2
            }
        }
        return null
    }

    fun fetchSightsListSync(district: String? = null): List<Sight> = runBlocking { fetchSightsList(district) }

    fun fetchSightDetailsSync(url: String): Sight? = runBlocking { fetchSightDetails(url) }

    suspend fun fetchSightsList(district: String? = null): List<Sight> = withContext(limitedIO) {
        val doc = fetchHtmlWithRetry("/tourguide/taiwan/keelungcity/") ?: return@withContext emptyList()

        val cssQuery = if (district == null) "#guide-point ul a" else "h4:contains(${district}) + ul a"
        val links = doc.select(cssQuery)

        log.debug("Found {} sight links", links.size)
        for (link in links) log.debug(" - [{}]({})", link.text(), link?.attr("href"))

        links.map { li ->
            async {
                val url = li.attr("href")
                fetchSightDetails(url)
            }
        }.awaitAll().filterNotNull()

    }

    suspend fun fetchSightDetails(url: String): Sight? = withContext(limitedIO) {
        val doc = fetchHtmlWithRetry(url) ?: return@withContext null

        val address = doc.selectFirst(".address a")
        val img = doc.selectFirst("#galleria .swiper-slide img")

        Sight(
            sightId = CrawlerUtils.getIDFromURL(url),
            name = doc.selectFirst("h1 > span")?.text(),
            city = doc.selectFirst(".breadcrumb .bc_li:nth-last-child(2) a")?.text(),
            district = doc.selectFirst(".breadcrumb .bc_last a")?.text(),
            category = doc.selectFirst(".point_type > span:nth-child(2)")?.text(),
            description = CrawlerUtils.getTextFromElement(doc.selectFirst(".text")),
            address = address?.text(),
            mapUrl = address?.attr("href"),
            photoUrl = img?.attr("data-src") ?: img?.attr("src"),
            sourceUrl = BASE_URL + url
        )
    }

}
